### ISSUE TITLE
Adding a stop check script during Deployment/Update phase 

### ISSUE THREAD

**Issue by aleroot** [aws/amazon-ecs-agent#1622]
### Summary

Since I have a web application that is also running some background tasks I would need to have a way to check if these tasks are finished before shutting down the task and the related docker container during a deployment/update phase.

### Description

Basically at the moment we we have a group of EC2 instances running docker(NO ECS) in which we upgrade to a new version of our web app, starting up a group of new instances with the new version of the docker image(with the updated web app), then after the new version is up and running we attach new instances to the load balancer and detach the old ones. After this migration phase we manually check if these old instances have completed all the pending tasks(we run bash script that return the number of pending tasks), before shutting them down(we shut down just when the count of the pending background tasks is equal to zero).

Is there a way to automate this process using ECS? 
Basically what we need is a feature that allows us to add a script to perform on the container at the moment of the stop task and that control the shutdown process of the task, for example, if the bash script return 0 then it is ok to stop the task, while if greater than zero run the script again after 30 seconds keeping the task in DRAINING state, not accepting new connections from load balancer, but just waiting for termination...

Another possible approach that we could embrace is adding termination timeout of the task that is longer than the actual 30 seconds, we would need something like 15 minutes...

**Comment by haikuoliu** [comment]
Hi @aleroot would [service update](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/update-service.html) help for your case? ECS Service can be used with load balancer as well.

**Comment by aleroot** [comment]
Yes, I'll update the service, but how can I make the task that is stopping wait until all background tasks are finished ?

I would need something like Kubernetes PreStop hook : https://kubernetes.io/docs/concepts/containers/container-lifecycle-hooks/

**Comment by haikuoliu** [comment]
Hi @aleroot, we don't support hooks right now, is #720 something what you want?

Thanks,
Haikuo

**Comment by aleroot** [comment]
What I need is a way to slow down the shutdown of a task, that is something doable on Kubernetes using the PreStop hook.

**Comment by adnxn** [comment]
@aleroot: I'm going to tag this as a feature request for now. We'll update the issue as we make progress on this feature. In the meantime, feel free to submit a design proposal or a more detailed use case of your requirements to help guide the design.
